Add L2 announcement IPv6 support

msune opened the pull request:
This patchset adds L2 announcement IPv6 support.

Patchset:

```
23c6929a58 docs/l2_announcements: remove IPv6 limitation
08019f3ef8 pkg: add support for L2 IPv6 announcements
db22a4c2d4 pkg/datapath: garp->gneigh and implement ND adv.
234e1b4a76 pkg: add v6 L2 responder maps
98ee4c6ff3 Add mdlayher/ndp go package
71721fa802 bpf/test: add L2 announce IPv6 unit test
749f591e3f bpf: Add support for IPv6 L2 announcements
aa2dc65eec bpf/test/pktgen: add IPv6 {svc,ext}_one addresses
623417f1c4 bpf: rescope IPv6 MC helpers (sol. addr)
```

Depends on #39574 and #39579. Related/implements #34983.

```release-note
Add L2 announcement IPv6 support 
```

msune committed:
bpf: Add support for IPv6 L2 announcements

This patch adds supports in bpf_host for IPv6 L2 announcements.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

msune committed:
pkg: add support for L2 IPv6 announcements

This commit adds support for IPv6 L2 announcements by adding the
necessary logic in the l2responder/ cell using the gratuitous ND
functionality now part of the gneigh cell.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

msune committed:
bpf/test/pktgen: add v6_svc_one address

Define v6_svc_one address required by the V6 l2_announce test.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

msune committed:
bpf/test: add L2 announce IPv6 unit test

This commit adds tc_l2_announce6 unit test, to cover IPv6 support
for L2 announce.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

msune committed:
Add mdlayher/ndp go package

This commit adds mdlayher/ndp golang package to implement support
for sending Gratuitous Neighbour packets.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

msune committed:
pkg: add v6 L2 responder map

This commit adds v6 L2 responder map.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

msune committed:
pkg/datapath: garp->gneigh and implement ND adv.

This commit:

* Renames garp cell to gneigh to be more generic.
* Adds support to gneigh for sending gratuitous Neighbour Advertisement
  messages.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

msune committed:
docs/l2_announcements: remove IPv6 limitation

On l2-announcements.rst:

* Remove IPv6 limitation.
* Adapt text to include NDP where only ARP was mentioned.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

msune committed:
bpf: rescope IPv6 MC helpers (sol. addr)

MC helpers are only really used for Solicitation addresses, therefore
make narrow the scope to make them more useful. Adapted tests.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

rgo3 commented on bpf/node_config.h:
```diff
@@ -147,6 +147,7 @@
 #define SRV6_POLICY_MAP_SIZE 16384
 #define SRV6_SID_MAP_SIZE 16384
 #define L2_RESPONDER_MAP4_SIZE 4096
+#define L2_RESPONDER_MAP6_SIZE 4096
```
Could map size macros already be configured through the runtime config mechanism (`DECLARE_CONFIG`)? If yes, this might be a nice place to do it and add a separate commit for the v4 map maybe? cc @dylandreimerink

dylandreimerink replied:
No, we cannot use `DECLARE_CONFIG`/`CONFIG` for these. Long term we will just set a default value in C, no more macros and any modifications will happen purely in Go. I think that falls under https://github.com/cilium/cilium/issues/39691

rgo3 reviewed (approved):
Loader changes LGTM!

HadrienPatte reviewed (approved):
vendor LGTM 👍

qmonnet commented on Documentation/network/l2-announcements.rst:
```diff
@@ -570,10 +568,10 @@ L2 Pod Announcements
 
 L2 Pod Announcements announce Pod IP addresses on the L2 network using
-Gratuitous ARP replies. When enabled, the node transmits Gratuitous ARP
-replies for every locally created pod, on the configured network
-interface(s). This feature is enabled separately from the above L2
-announcements feature.
+Gratuitous ARP replies / Neighbor Discovery Advertisements. When enabled, the
+node transmits Gratuitous ARP replies / NDP Advertisements for every locally
+created pod, on the configured network interface(s). This feature is enabled
+separately from the above L2 announcements feature.
```
This section also has a note at the end that says:

> Since this feature has no IPv6 support yet, only ARP messages are sent, no Unsolicited Neighbor Advertisements are sent.

Can you please update it?

msune replied:
Good catch. Fixed thx.

qmonnet commented on pkg/datapath/gneigh/cells.go:
```diff
@@ -28,10 +28,10 @@ func (def Config) Enabled() bool {
 
 func (def Config) Flags(flags *pflag.FlagSet) {
-	flags.String(L2PodAnnouncementsInterface, def.L2PodAnnouncementsInterface, "Interface used for sending gratuitous arp messages")
+	flags.String(L2PodAnnouncementsInterface, def.L2PodAnnouncementsInterface, "Interface used for sending gratuitous arp and nd messages")
```
Nit: why not fix the case?

```suggestion
	flags.String(L2PodAnnouncementsInterface, def.L2PodAnnouncementsInterface, "Interface used for sending gratuitous ARP and ND messages")
```

msune replied:
Done, and also used `NDP` instead of `ND` for consistency.

qmonnet commented on pkg/datapath/gneigh/cells.go:
```diff
@@ -28,10 +28,10 @@ func (def Config) Enabled() bool {
 	flags.MarkDeprecated(L2PodAnnouncementsInterface, "use --"+L2PodAnnouncementsInterfacePattern+" instead")
-	flags.String(L2PodAnnouncementsInterfacePattern, def.L2PodAnnouncementsInterfacePattern, "Regex matching interfaces used for sending gratuitous arp messages")
-	flags.Bool(EnableL2PodAnnouncements, def.EnableL2PodAnnouncements, "Enable announcing Pod IPs with Gratuitous ARP")
+	flags.String(L2PodAnnouncementsInterfacePattern, def.L2PodAnnouncementsInterfacePattern, "Regex matching interfaces used for sending gratuitous arp and nd messages")
```
```suggestion
	flags.String(L2PodAnnouncementsInterfacePattern, def.L2PodAnnouncementsInterfacePattern, "Regex matching interfaces used for sending gratuitous ARP and ND messages")
```

msune replied:
Done, and also used `NDP` instead of `ND` for consistency.

qmonnet reviewed (changes requested):
Great work!

Please fix the remaining note in the docs (see below).

msune committed:
pkg/l2_responder: continue if gARP/NDP fails

Prior to this commit if a gratuitous ARP/NDP was unable to be sent,
the L2 announce map (v4/v6) was not populated with the entry,
leading to an inconsistency. This was not a big deal as _generally_
nothing should prevent us to send these packets.

Change the behaviour to Warn and continue.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

qmonnet reviewed (approved):
Thank you :slightly_smiling_face:

joestringer commented on bpf/lib/ipv6.h:
```diff
@@ -195,24 +195,25 @@ static __always_inline bool ipv6_addr_equals(const union v6addr *a,
 	mac->addr[0] = 0x33;
 	mac->addr[1] = 0x33;
-	memcpy((__u8 *)mac + 2, (__u8 *)addr + 12, 4);
+	mac->addr[2] = 0xFF;
+	memcpy((__u8 *)mac + 3, (__u8 *)addr + 13, 3);
```
This commit doesn't describe why the change for the third octet here, which makes me curious what impact this specific change has and why it's "more useful" to do it this way?

I went down a small rabbit hole trying to find an authoritative source on this and I came up empty. The previous implementation seems to [follow RFC2464](https://datatracker.ietf.org/doc/html/rfc2464#section-7). Wikipedia has an entry that affirms the new version, but the article only references a [presentation related to an expired internet draft](https://www.ietf.org/proceedings/65/slides/16ng-3/sld5.htm) which [doesn't even specify this prefix](https://datatracker.ietf.org/doc/draft-neves-rt-usage-80216-problem-statement/). Wondering if maybe this is one of those things that's deep in some IEEE standard somewhere.

I guess that realistically speaking if this is in fact the standard and everyone is transmitting such packets with `33:33:FF:XX:XX:XX` then this change makes no practical difference. In that case it's hardly something to hold the PR back on. I just figure that if we're adhering to some standard then it's worthwhile to confirm the basis for the change.

msune replied:
>I guess that realistically speaking if this is in fact the standard and everyone is transmitting such packets with 33:33:FF:XX:XX:XX

Not exactly.

The MAC address we need is created from the encoding of the solicited-node MC **IPv6 address**, which is a "pre-defined" multicast address defined in [RFC4291, section 2.7.1 "pre-defined multicast addresses"](https://datatracker.ietf.org/doc/html/rfc4291#section-2.7.1). The RFC states that the 4th (least significant) byte  must be `FF` and the 3 LSBs must be the target IP's 24 least significant bits:

```
      Solicited-Node Address:  FF02:0:0:0:0:1:FFXX:XXXX
```

This is what results in _all_ solicited node **MAC** addresses start with `33:33:FF`.

Currently we are only supporting NDP in the BPF datapath. If I were to have left the MAC helper as it was - note, not used by anyone else - for the NDP case the caller would have had to:

1. Get the solicited-node IPv6 address (in stack)
2. Call the MAC helper

By narrowing down the helper [in this commit](https://github.com/cilium/cilium/pull/39648/commits/2776588e454136240d72de2c6d0529841dd5353d), from `ipv6_mc_mac_set()` -> `ipv6_sol_mc_addr_set()` I avoid having to do 1). If at some point we need `ipv6_mc_addr_set()`, we can define it as an additional helper.

---

More info:

This is in addition to other pre-defined IPV6 addresses, such as the [v6 MC addresses](https://datatracker.ietf.org/doc/html/rfc4291#section-2.7):

```
      Reserved Multicast Addresses:   FF00:0:0:0:0:0:0:0
                                      FF01:0:0:0:0:0:0:0
                                      FF02:0:0:0:0:0:0:0
                                      FF03:0:0:0:0:0:0:0
                                      FF04:0:0:0:0:0:0:0
                                      FF05:0:0:0:0:0:0:0
                                      FF06:0:0:0:0:0:0:0
                                      FF07:0:0:0:0:0:0:0
                                      FF08:0:0:0:0:0:0:0
                                      FF09:0:0:0:0:0:0:0
                                      FF0A:0:0:0:0:0:0:0
                                      FF0B:0:0:0:0:0:0:0
                                      FF0C:0:0:0:0:0:0:0
                                      FF0D:0:0:0:0:0:0:0
                                      FF0E:0:0:0:0:0:0:0
                                      FF0F:0:0:0:0:0:0:0

   The above multicast addresses are reserved and shall never be
   assigned to any multicast group.

      All Nodes Addresses:    FF01:0:0:0:0:0:0:1
                              FF02:0:0:0:0:0:0:1

   The above multicast addresses identify the group of all IPv6 nodes,
   within scope 1 (interface-local) or 2 (link-local).

      All Routers Addresses:   FF01:0:0:0:0:0:0:2
                               FF02:0:0:0:0:0:0:2
                               FF05:0:0:0:0:0:0:2
```

The mapping you are describing, so putting the whole 4 LS bytes in the destination MAC address is only used for "regular" IPv6 MC traffic.

joestringer replied:
Ah great. This was just a bit more nuanced than I expected. The commit message for this commit could benefit from having this detail included, as this is not super obvious. But that's not a big deal. Thanks for taking the time to explain this to me, I'm quite unfamiliar with this area :)

joestringer commented on bpf/bpf_host.c:
```diff
@@ -1083,8 +1111,22 @@ do_netdev(struct __ctx_buff *ctx, __u16 proto, __u32 __maybe_unused identity,
+			ret = handle_l2_announcement(ctx, ip6);
+			if (ret != CTX_ACT_OK)
+				break;
+			/* Verifier invalidates ip6 for some reason.. sigh*/
+			if (!revalidate_data_pull(ctx, &data, &data_end, &ip6))
+				return send_drop_notify_error(ctx, identity,
+							      DROP_INVALID,
+							      METRIC_INGRESS);
```
nit: Might be related with the change of the tail in `icmp6_ndisc_adv_addopt()`. I'd be a little curious whether it's enough to revalidate in that conditional under `icmp6_send_ndisc_adv()` to reduce the number of conditions in which we have to do this extra call.

No need to change the logic in this PR.

msune replied:
>Might be related with the change of the tail in icmp6_ndisc_adv_addopt()

I vaguely recall this issue even before `icmp6_ndisc_adv_addopt()`. I would suggest to leave it like this, maybe someone else can have a look into _why_ the verifier invalidates `ip6`.

joestringer commented on bpf/lib/icmp6.h:
```diff
@@ -524,4 +524,42 @@ icmp6_host_handle(struct __ctx_buff *ctx, int l4_off, __s8 *ext_err, bool handle
+	struct icmp6hdr *icmp;
+	int l3_off = (int)((__u8 *)ip6 - (__u8 *)ctx_data(ctx));
+	int l4_off = ipv6_hdrlen(ctx, &nexthdr);
+	struct ethhdr *eth = (void *)(long)ctx->data;
```
nit for consistency and simplification:
```suggestion
	struct ethhdr *eth = ctx_data(ctx);
```

msune replied:
Good catch. Fixed.

joestringer commented on bpf/lib/icmp6.h:
```diff
@@ -524,4 +524,42 @@ icmp6_host_handle(struct __ctx_buff *ctx, int l4_off, __s8 *ext_err, bool handle
+	struct ethhdr *eth = (void *)(long)ctx->data;
+	union macaddr *dmac;
+
+	if ((void *)eth + ETH_HLEN > (void *)(long)ctx->data_end)
```
nit:
```suggestion
	if ((void *)eth + ETH_HLEN > ctx_data_end(ctx))
```

msune replied:
Good catch. Fixed.

joestringer commented on bpf/bpf_host.c:
```diff
@@ -1030,22 +1032,48 @@ static __always_inline int handle_l2_announcement(struct __ctx_buff *ctx)
+		struct l2_responder_v6_key key6;
+		int l3_off;
+
+		if (!icmp6_ndisc_validate(ctx, ip6, &mac, &smac, &sip6, &tip6))
```
`smac` and `sip6` are unused.

```suggestion
		if (!icmp6_ndisc_validate(ctx, ip6, &dmac, &tip6))
```

msune replied:
`mac` and `sip`  are very much used, and are output params of the function: https://github.com/cilium/cilium/blob/369f79124da775f66c45f406da0c4c2c2ec53b67/bpf/lib/icmp6.h#L554-L557

I will fix the incorrect `const` in `smac` (cast in `eth_load_saddr()` masks this error).

msune replied:
Damn.. this seems a bug in cocci:

https://github.com/cilium/cilium/actions/runs/16191362202/job/45707705664?pr=39648#step:4:15

msune replied:
(Reverted to add the - incorrect - `const` statement for `smac`)

joestringer replied:
Am I missing something? `handle_l2_announcement()` does not return `smac` or `sip6`. Yes, `icmp6_ndisc_validate()` returns these as output parameters. For the `arp_validate()` case, the equivalent source address parameters are subsequently passed to `arp_respond()`. But for this new IPv6 case, these are stack variables that are passed to `icmp6_ndisc_validate()` which returns them as output parameters, then the output parameters are not used later in `handle_l2_announcement()`.

joestringer replied:
<img width="795" height="1283" alt="image" src="https://github.com/user-attachments/assets/82f458d6-c07c-4202-a722-11f3a2b26dd3" />

msune replied:
Oh! Sorry, I didn't look into the caller :sweat_smile: 

Ok, `smac` removed from the param list, and `sip6` removed from param and from stack :rocket:

joestringer commented on bpf/lib/icmp6.h:
```diff
@@ -524,4 +524,42 @@ icmp6_host_handle(struct __ctx_buff *ctx, int l4_off, __s8 *ext_err, bool handle
+	ipv6_load_saddr(ctx, l3_off, sip);
+	*tip = *(union v6addr *)(icmp + 1);
+
+	if (!ipv6_is_sol_mc_mac(tip, dmac) && eth_addrcmp(dmac, iface_mac) != 0)
```
nit: The first condition here ensures that `tip` and `dmac` share the last three octets and that `dmac` has the first three octets set to `33:33:FF`. At a glance it seems correct to do the second check, but at the same time I don't see why anyone would set such a multicast mac as the interface mac. Do you think that part is important to either include or exclude?

joestringer replied:
Maybe put another way, having such an interface MAC seems like something we could detect earlier in userspace as a sanity check and then drop the extra sanity check here :shrug:

msune replied:
I am not sure I follow here. `tip` and `dmac` are coming from the Neighbour Discovery Solicitation original packet. What this statement does is to validate that the destination mac of the NS is _either_ `iface_mac` (targeted NS, so a refresh in the neighbour state) or the v6 MC mac address (solicited node MC address).

This is in accordance to [RFC4861, section 4.3](https://datatracker.ietf.org/doc/html/rfc4861#section-4.3) in which for a NS packet:

```
Destination Address
                     Either the solicited-node multicast address
                     corresponding to the target address, or the target
                     address.
```

In other words, this would discard any packet that might target a valid SVC IP but going to the wrong MAC addr (e.g. another node's one, or a bogus bcast_addr for instance). Note that, while the MC mac address is meant to be used to "early filter" packets not destined to the node, in order to compose the MC MAC address you need the target IP, which in order to be properly implemented (essentialy for ND proxying) you should get it from ICMPv6 target IP, and not from IP's dst_ip.

So, I don't think we can "precompute" anything.

joestringer replied:
Ah right my assumptions about the packet content going in were incorrect. Your explanation makes sense to me :+1:

joestringer commented on pkg/maps/l2v6respondermap/l2_responder_map6.go:
```diff
@@ -0,0 +1,135 @@
+
+var Cell = cell.Provide(NewMap)
+
+type Map interface {
+	Create(ip netip.Addr, ifIndex uint32) error
+	Lookup(ip netip.Addr, ifIndex uint32) (*l2respondermap.L2ResponderStats, error)
+	Delete(ip netip.Addr, ifIndex uint32) error
+	IterateWithCallback(cb IterateCallback) error
+}
```
nit: Seems like we could avoid defining a different interface here and share the one with the l2respondermap.

msune replied:
I will address this in the other comment about a common map, as it boils down - I believe - to the same.

joestringer commented on pkg/maps/l2v6respondermap/l2_responder_map6.go:
```diff
@@ -0,0 +1,135 @@
+
+// L2V6ResponderKey implements the bpf.MapKey interface.
+//
+// Must be in sync with struct l2_responder_v4_key in <bpf/lib/maps.h>
```
There is no such file `bpf/lib/maps.h`. Also, this should be in sync with the v6 structure, not `l2_responder_v4_key`.

msune replied:
The `bpf/lib/maps.h` comes from the v4 comment: https://github.com/cilium/cilium/blob/main/pkg/maps/l2respondermap/l2_responder_map4.go#L136

Fixed the v4 -> v6, and fixed both comments.

joestringer commented on pkg/maps/l2v6respondermap/l2_responder_map6.go:
```diff
@@ -0,0 +1,135 @@
+//
+// Must be in sync with struct l2_responder_v4_key in <bpf/lib/maps.h>
+type L2V6ResponderKey struct {
+	IP      types.IPv6 `align:"address"`
```
The corresponding definition in `bpf/lib/l2_responder.h` defined this field with the name `ip6`, which tells me that the align checker is not validating that the Go and C structures have equivalent alignment (or maybe something has changed there since I last touched alignchecker?).
```suggestion
	IP      types.IPv6 `align:"ip6"`
```

joestringer replied:
@cilium/sig-datapath we might want to consider looking into `bpf/bpf_alignchecker.c` to see whether we're still ensuring alignment between Go and C versions of these structures.

msune replied:
FWIW, in `main` v4 is also not matching (`ip4` vs `ip`):

https://github.com/cilium/cilium/blob/main/bpf/lib/l2_responder.h#L11
https://github.com/cilium/cilium/blob/main/pkg/maps/l2respondermap/l2_responder_map4.go#L119

I will fix it as part of adding the map commit, but I think the alignchecker discussion should be stubbed somewhere else.

joestringer replied:
Yep makes sense. I'll mark this thread as "resolved" as it doesn't need to be addressed in this PR. I just think that the codeowner group should consider how we're ensuring proper alignment of these fields going forward; it seems like somewhere along the way we stopped enforcing the same mechanism for alignment. I know that early on in the project, it was semi-common for alignment to differ and this wasted a lot of contributor time to figure out, which was why we put the automated check in place. I guess that context was not passed on to newer reviewers as they joined the group.

joestringer commented on pkg/maps/l2v6respondermap/l2_responder_map6_fake.go:
```diff
@@ -0,0 +1,45 @@
+// SPDX-License-Identifier: Apache-2.0
```
nit, though I see the existing logic follows the same pattern: It'd be nice to avoid having fake and testing implementations exposed in the production packages. We could consider putting this under `pkg/testutils`. Extra points to share the same fake implementation for both maps since it ultimately doesn't matter what the underlying storage key is when the interface is the same for both fake implementations:

```
$ diff --unified pkg/maps/l2respondermap/l2_responder_map4_fake.go pkg/maps/l2v6respondermap/l2_responder_map6_fake.go
--- pkg/maps/l2respondermap/l2_responder_map4_fake.go   2024-09-06 08:12:48.242549701 -0700
+++ pkg/maps/l2v6respondermap/l2_responder_map6_fake.go 2025-07-09 13:26:23.355776099 -0700
@@ -1,27 +1,29 @@
 // SPDX-License-Identifier: Apache-2.0
 // Copyright Authors of Cilium
 
-package l2respondermap
+package l2v6respondermap
 
 import (
        "net/netip"
+
+       "github.com/cilium/cilium/pkg/maps/l2respondermap"
 )
 
 func NewFakeMap() Map {
-       return &fakeMap{entries: make(map[L2ResponderKey]L2ResponderStats)}
+       return &fakeMap{entries: make(map[L2V6ResponderKey]l2respondermap.L2ResponderStats)}
 }
 
 type fakeMap struct {
-       entries map[L2ResponderKey]L2ResponderStats
+       entries map[L2V6ResponderKey]l2respondermap.L2ResponderStats
 }
 
 func (fm *fakeMap) Create(ip netip.Addr, ifIndex uint32) error {
-       fm.entries[newL2ResponderKey(ip, ifIndex)] = L2ResponderStats{}
+       fm.entries[newL2V6ResponderKey(ip, ifIndex)] = l2respondermap.L2ResponderStats{}
        return nil
 }
 
-func (fm *fakeMap) Lookup(ip netip.Addr, ifIndex uint32) (*L2ResponderStats, error) {
-       entry, found := fm.entries[newL2ResponderKey(ip, ifIndex)]
+func (fm *fakeMap) Lookup(ip netip.Addr, ifIndex uint32) (*l2respondermap.L2ResponderStats, error) {
+       entry, found := fm.entries[newL2V6ResponderKey(ip, ifIndex)]
        if found {
                return &entry, nil
        }
@@ -30,7 +32,7 @@
 }
 
 func (fm *fakeMap) Delete(ip netip.Addr, ifIndex uint32) error {
-       delete(fm.entries, newL2ResponderKey(ip, ifIndex))
+       delete(fm.entries, newL2V6ResponderKey(ip, ifIndex))
        return nil
 }

```

msune replied:
I am going to be honest here; I don't have much Golang experience, so I was following the pattern of the v4 part. I would suggest to tackle this in a separate PR, given that a) this PR is _long_ overdue b) it follows the existing code structure and pattern.

cc @dylandreimerink

dylandreimerink replied:
> nit, though I see the existing logic follows the same pattern: It'd be nice to avoid having fake and testing implementations exposed in the production packages

Like you said, this is sort of grandfathered in. Ideally we would go the loadbalancer approach (pinned map in production, unpinned maps during testing on linux, fake maps for non-linux). Doing so requires refactoring quite a bit. I second the idea of doing cleanup / improvement of tests in a separate PR.

joestringer replied:
I agree, this is not critical enough to require addressing this point prior to merge. This can be a subsequent code maintenance question to improve for code owners of this area.

tommyp1ckles replied:
Just as a side discussion, I'm always a bit weary of mocking bpf maps instead of just creating them in privileged tests. A Go map is probably only a rough estimation of actual bpf maps.

joestringer replied:
I can understand the weariness there, but at the same time I think that for control plane testing it would be ideal to avoid hard dependencies on BPF maps. I'd like if tests like the ones using these maps can be run on mac OS for instance, since there is a subset of the Cilium developer community on that platform.

joestringer commented on pkg/datapath/gneigh/processor.go:
```diff
@@ -147,36 +152,62 @@ func (gp *processor) updateInterfaces() (<-chan struct{}, error) {
-	if newIP.IsUnspecified() {
-		return
+	if ep.IPv4.IsUnspecified() && ep.IPv6.IsUnspecified() {
+		return //Bug
```
nit: If this is a bug, we could consider emitting an error/warning message. CI would catch that and highlight.

msune replied:
OK, `Warn` trace added.

joestringer commented on pkg/datapath/gneigh/processor.go:
```diff
@@ -147,36 +152,62 @@ func (gp *processor) updateInterfaces() (<-chan struct{}, error) {
-				logfields.IPAddr, newIP,
-			)
-		}
+		gp.send(ep, newIPs.IPv4, iface)
+		gp.send(ep, newIPs.IPv6, iface)
```
What happens in single-stack case if one of these IPs is nil?

msune replied:
Not an expert on Golang, but to the best of my knowledge, neither `IPv4` nor `IPv6` can be nil, as they come from the `ep` [here](https://github.com/cilium/cilium/blob/369f79124da775f66c45f406da0c4c2c2ec53b67/pkg/datapath/gneigh/processor.go#L196-L199) and endpoint defines it as an object, not a reference (see [this](https://github.com/cilium/cilium/blob/main/pkg/endpoint/endpoint.go#L241-L251)).

Addresses might be unsued (`Unspecified`), which is then handled in the `send()` routine:

https://github.com/cilium/cilium/blob/369f79124da775f66c45f406da0c4c2c2ec53b67/pkg/datapath/gneigh/processor.go#L156-L171

@dylandreimerink I would appreciate another pair of eyes on this one.

dylandreimerink replied:
> What happens in single-stack case if one of these IPs is nil?

I am not 100% sure what the expected value of these fields are when their respective protocol is disabled. I would assume they are the default zero-initialized value of `netip.Addr`. 

We check `IsUnspecified` in the `send` method:
```
if ip.IsUnspecified() {
	return
}
```
Which feels right. However the [docs](https://pkg.go.dev/net/netip#Addr.IsUnspecified) say:
> Note that the zero [Addr](https://pkg.go.dev/net/netip#Addr) is not an unspecified address.

It seems that `IsValid` is what we want to check instead:
> IsValid reports whether the [Addr](https://pkg.go.dev/net/netip#Addr) is an initialized address (not the zero Addr).

However, we should probably double check the behavior of endpoint. For all I know we use some odd placeholder value instead to signal not having an IP.

msune replied:
Changed to use `!isValid()` instead:

* here: https://github.com/cilium/cilium/pull/39648/commits/d479a564781fac73145a227bc471fa7a0d0e2e16#diff-2c7f1cead6899ddf0bfc4adce48aa43c487b9d2cbfac15f64192b4bcaeb532f6R156
* And here: https://github.com/cilium/cilium/pull/39648/commits/d479a564781fac73145a227bc471fa7a0d0e2e16#diff-2c7f1cead6899ddf0bfc4adce48aa43c487b9d2cbfac15f64192b4bcaeb532f6R192

I've looked into the code, and this is consistent with other parts. They use `!isValid()` in all cases. Squashed as part of the gneigh commit.

joestringer replied:
I miscommunicated this when referring to the addresses as "nil", indeed `IsValid()` is the way I'd expect this to be handled. Looks good, thanks!

joestringer commented on pkg/datapath/l2responder/l2responder.go:
```diff
@@ -116,36 +118,46 @@ func (p *l2ResponderReconciler) cycle(
 			return nil
 		}
 
-		err = garpOnNewEntry(arMap, p.params.GARPSender, e.IP, idx)
+		if e.IP.Is6() {
+			err = gneighOnNewEntry(ndMap, p.params.GNeighSender, e.IP, idx)
+		} else {
+			err = garpOnNewEntry(arMap, p.params.GNeighSender, e.IP, idx)
+		}
 		if err != nil {
-			return err
+			log.Warn("Unable to send gratuitous ARP/NDP. Continuing...", logfields.Error, err)
```
What does this change mean for retries?

msune replied:
Not sure I understand the question; in short, there aren't garp/gndp retries _today_. This is to invalidate ARP/NDP cache of the neighbours, so it is only sent once. 

This change comes from this https://github.com/cilium/cilium/pull/39648#issuecomment-3012471545, as explained in [this commit](https://github.com/cilium/cilium/pull/39648/commits/369f79124da775f66c45f406da0c4c2c2ec53b67)

tommyp1ckles replied:
Think this makes sense to me, the process function returns an error and will [break early on one of the changes if it is not null](https://github.com/cilium/cilium/pull/39648/commits/69f5f02af0a0376706d575b4ffbf42f0581b3a55#diff-e399ceade16cec15475fccfa6aeb3d7d2b79bdba2463acb76242efd78edde9fdR172).

@msune In this case the idea is that; if the GARP step fails we don't want to block subsequent changes from having their changes processed, otherwise we might be indefinitely skipping a change we could have reconciled.

One improvement we could potentially do here is to use a cell.Health handle to emit a degraded state if the GARP fails?

joestringer commented on pkg/datapath/l2responder/l2responder.go:
```diff
@@ -231,21 +250,35 @@ func (p *l2ResponderReconciler) fullReconciliation(txn statedb.ReadTxn) (err err
+			toDelete6 = append(toDelete6, key)
+			return
+		}
+		e.satisfied = true
```
This statement is a noop. Same problem for the IPv4 version. `e` is declared on the stack, copied as a struct from the map, then this updates the stack version and then it falls out of scope on the next line to be thrown away.

msune replied:
Uh, good catch. @dylandreimerink  is this some left over from a previous iteration of the original v4 support in  l2_responder, or is it somehow a bug.

Not sure what `satisfied` was supposed to do here - rationale - so I guess it's safe to delete the [struct field](https://github.com/cilium/cilium/blob/main/pkg/datapath/l2responder/l2responder.go#L199) and all references, ifs() etc...

dylandreimerink replied:
The responder maps value is a counter that keeps track of the amount of times we have responded (mostly for debug purposes)

The purpose of `e.satisfied` is that when we loop over the entries later on we do not overwrite the value of existing entries, only create new entries.

So, since we modify a stack copy of `e` it means we reset the counters of the map value every time we reconcile. Adding `desiredMap6[*key] = e` should fix it.

Its not a critical issue, the counter isn't used for anything other than debugging at the moment.

msune replied:
>Adding desiredMap6[*key] = e should fix it.

Added in [this commit](https://github.com/cilium/cilium/pull/39648/commits/61237f6308858b19daeb90c7c063ef11fd3b2b94) so that it can eventually be backported (will not apply). Credited both of you.

msune replied:
@joestringer please reopen if you disagree with the fix. thx

joestringer replied:
Fix looks good :+1:

joestringer commented on pkg/datapath/l2responder/l2responder.go:
I found myself wanting for there to be a better IPv4 vs. IPv6 abstraction in this file. Much of the logic is directly duplicated with interleaving IPv4 vs. IPv6 statements. On one hand I felt that this made it easier to review because the corresponding IPv4 version was right above, but on the other hand the resulting code is a bit more stilted to read because you need to pick whether you care about IPv4 or IPv6 and then skip every second hunk.

Functionally it seems fine and it's not like we'll save effort for future IP protocols by building out a better abstraction. Maybe a bit of copying is better than introducing too many abstractions here. At least if there's a bug, we're very likely to consider how the bug may impact both IPv4 and IPv6 at the same time since the code is right there side-by-side.

msune replied:
I don't disagree. I don't think we will have v7/v8/v10 anytime soon, though :sweat: 

On a serious note, I made the same exact appreciation in the original RFC PR [here](https://github.com/cilium/cilium/pull/34983):

> Not particulary happy with the golang code under l2responder.go. It's a bit ugly with the if(v6) else ...; I am new to golang, and I haven't seen this pattern used in Maps, but would it make sense to have a common interface for both l2respondermaps?

So for me this ship has sailed long time ago. The purpose of an RFC PR is to address fundamental arch issues, so that we don't have to do major changes afterwards.

joestringer commented on pkg/maps/l2v6respondermap/l2_responder_map6.go:
Apologies if I missed it, but did you consider keeping one map for both protocols? Might help resource usage / map handling a little bit.

msune replied:
I did, yes. Several comments:

1. I was following - without knowing it, at the time - the recommendation [here](https://github.com/cilium/cilium/issues/28985#issuecomment-1795059665) ("new responder map") by @dylandreimerink 
2. This was implicitely asked in the RFC PR [here](https://github.com/cilium/cilium/pull/34983)
3. There is a small tradeoff in which, if you are not using v6, the l2_responder map (keys) would be 4x in memory. Not a big issue, though.

I will put emphasis in 2). Refactoring to have a single map _now_ is a major change. This PR has been floating around for a very very long time (arguably because of my lack of time), but still after:

a) in the RFC PR which this was asked I didn't get any feedback and
b) almost completed all reviews.

So, I don't think I can (nor want to) spend time trying to refactor this PR to use a single map. This can be a good first issue for someone, along with moving some of the BPF code from `bpf_host.h` to a dedicated header `lib/l2_announce.h` or alike.

dylandreimerink replied:
The benefit of having 1 map is slight, given they are global maps with only 4k entries. Combining at some point does not hurt. But I don't think it has to be a blocker

joestringer reviewed (commented):
Nice work! A bunch of my feedback is around little things like extra unneeded parameters, some ineffective statements, some ways we could simplify the code (eg avoid casts). There's some comments that reference the wrong structures or files that don't exist. All nice-to-have things but not critical; see below for details.

The most notable feedback below is around how single-stack is handled, since it didn't look like `pkg/datapath/gneigh/processor.go` is aware of which protocols are enabled. Seemed like this might cause errors to show up in agent logs for protocols that are disabled. But maybe I just missed some detail about how the disabled protocols are skipped over. 

I raised some other lesser questions like whether we really need another map or could reuse the same map for both address families, and also some ideas about simplifying / reusing Go code. See below for the threads. I think these are worth considering, but I'd be fine with merging this PR and then iterating on top.

tommyp1ckles commented on bpf/tests/pktgen.h:
```diff
@@ -79,6 +79,10 @@ volatile const __u8 v6_pod_one[] = v6_pod_one_addr;
 volatile const __u8 v6_pod_three[] = v6_pod_three_addr;
 
+/* IPv6 addresses for hosts, external to the cluster */
+volatile const __u8 v6_ext_one[] = {0xfd, 0x03, 0, 0, 0, 0, 0, 0,
```
Note: this [commit](https://github.com/cilium/cilium/commit/3ab063603825cd240ea2a7d5471a0c0e2fb241d6) recently introduced some similar fixtures, maybe there's overlap.

msune replied:
Good catch, thx. Problems of having long standing PRs. Now we have to merge this one finally :sweat_smile:

msune replied:
Btw I took advantage to define `v6_svc_one` with the same pattern addr + global var.

msune committed:
pkg/l2_responder: fix counters 0ed on sync

As part of #39648 review, a bug was disovered in which l2_responder
counters were 0ed during reconciliation.

Fix it by making sure desiredMap contains old entries with satisfied=true
so that counters are not 0ed, as per suggested by Dylan.

Reported-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Dylan Reimerink <dylan.reimerink@isovalent.com>
Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

joestringer reviewed (approved):
LGTM for my code owners. I followed up on one point from my prior review here: https://github.com/cilium/cilium/pull/39648/files#r2198302693 , but I think this is also OK to fix up after merging this.

tommyp1ckles commented on go.mod:
```diff
@@ -71,6 +71,7 @@ require (
 	github.com/mattn/go-shellwords v1.0.12
 	github.com/mdlayher/arp v0.0.0-20220512170110-6706a2966875
 	github.com/mdlayher/ethernet v0.0.0-20220221185849-529eae5b6118
+	github.com/mdlayher/ndp v1.1.0
```
Wonder if its worth updating the other mdlayer libraries so they're all on a v1.x version: https://github.com/cilium/cilium/blob/main/go.mod#L72-L73

msune replied:
It probably is, but I would defer this to a follow up PR.

tommyp1ckles commented on pkg/maps/l2v6respondermap/l2_responder_map6.go:
```diff
@@ -0,0 +1,133 @@
+	DefaultMaxEntries = 4096
+)
+
+var Cell = cell.Provide(NewMap)
```
nit: general convention is to put the exported Cell var in a file named `cell.go`

msune replied:
I was really following the same pattern as the [v4 implementation](https://github.com/cilium/cilium/blob/main/pkg/maps/l2respondermap/l2_responder_map4.go#L25). I would suggest to do a follow up refactor of both pieces of code.

tommyp1ckles commented on pkg/maps/l2v6respondermap/l2_responder_map6.go:
```diff
@@ -0,0 +1,133 @@
+// passing each key/value pair to the cb callback.
+func (m *l2V6ResponderMap) IterateWithCallback(cb IterateCallback) error {
+	return m.Map.IterateWithCallback(&L2V6ResponderKey{}, &l2respondermap.L2ResponderStats{},
+		func(k, v interface{}) {
```
nit `k, v interface{}` -> `k, v any'

msune replied:
Good catch.. this changes 2-3 months ago in the v4 impl. and went unnoticed (https://github.com/cilium/cilium/commit/77edeb8377604e92a570316c93820161db23feb4). Fixed and squashed in the map commit.

tommyp1ckles reviewed (approved):
Just some minor stuff, but the changes look good to me.
